libs/libcommon/include/common/logger_useful.h

```diff
@@ -60,61 +50,58 @@ std::string toCheckedFmtStr(const S & format, const Ignored &, Args &&... args)
+    {                                                                       \
+        Poco::Message poco_message(                                         \
+            /*source*/ (logger)->name(),                                    \
+            /*text*/ (message),                                             \
+            /*prio*/ (PRIORITY),                                            \
+            /*file*/ &__FILE__[LogFmtDetails::getFileNameOffset(__FILE__)], \
+            /*line*/ __LINE__);                                             \
+        (logger)->log(poco_message);                                        \
     } while (false)
 
-#define LOG_FMT_TRACE(logger, ...) LOG_FMT_IMPL(logger, Poco::Message::PRIO_TRACE, __VA_ARGS__)
-#define LOG_FMT_DEBUG(logger, ...) LOG_FMT_IMPL(logger, Poco::Message::PRIO_DEBUG, __VA_ARGS__)
-#define LOG_FMT_INFO(logger, ...) LOG_FMT_IMPL(logger, Poco::Message::PRIO_INFORMATION, __VA_ARGS__)
-#define LOG_FMT_WARNING(logger, ...) LOG_FMT_IMPL(logger, Poco::Message::PRIO_WARNING, __VA_ARGS__)
-#define LOG_FMT_ERROR(logger, ...) LOG_FMT_IMPL(logger, Poco::Message::PRIO_ERROR, __VA_ARGS__)
-#define LOG_FMT_FATAL(logger, ...) LOG_FMT_IMPL(logger, Poco::Message::PRIO_FATAL, __VA_ARGS__)
+
+#define LOG_IMPL_0(logger, PRIORITY, message)        \
+    do                                               \
+    {                                                \
+        if ((logger)->is((PRIORITY)))                \
+            LOG_INTERNAL(logger, PRIORITY, message); \
+    } while (false)
+
+#define LOG_IMPL_1(logger, PRIORITY, fmt_str, ...)                                            \
+    do                                                                                        \
+    {                                                                                         \
+        if ((logger)->is((PRIORITY)))                                                         \
+        {                                                                                     \
+            auto _message = LogFmtDetails::toCheckedFmtStr(FMT_STRING(fmt_str), __VA_ARGS__); \
+            LOG_INTERNAL(logger, PRIORITY, _message);                                         \
+        }                                                                                     \
+    } while (false)
+
+#define LOG_IMPL_CHOSER(...) TF_GET_29TH_ARG(__VA_ARGS__, LOG_IMPL_1, LOG_IMPL_1, LOG_IMPL_1, LOG_IMPL_1, LOG_IMPL_1, LOG_IMPL_1, LOG_IMPL_1, LOG_IMPL_1, LOG_IMPL_1, LOG_IMPL_1, LOG_IMPL_1, LOG_IMPL_1, LOG_IMPL_1, LOG_IMPL_1, LOG_IMPL_1, LOG_IMPL_1, LOG_IMPL_1, LOG_IMPL_1, LOG_IMPL_1, LOG_IMPL_1, LOG_IMPL_1, LOG_IMPL_1, LOG_IMPL_1, LOG_IMPL_1, LOG_IMPL_1, LOG_IMPL_1, LOG_IMPL_1, LOG_IMPL_0)
+
+// clang-format off
+#define LOG_IMPL(logger, PRIORITY, ...) LOG_IMPL_CHOSER(__VA_ARGS__)(logger, PRIORITY, __VA_ARGS__)
+// clang-format on
+
+#define LOG_TRACE(logger, ...) LOG_IMPL(logger, Poco::Message::PRIO_TRACE, __VA_ARGS__)
```

Comment:
According to the doc:

> A [Message](https://docs.pocoproject.org/current/Poco.Message.html) can also contain any number of named parameters that contain additional information about the event that caused the message.


Seems that we can even consider supporting named arguments in future?

Reply:
Does "named arguments" means a behavior like this?
```
int val = 12345;
LOG_DEBUG(log, "log some value"; val);
// generate a log "log some value [val=12345]"
```

As `Poco::Message` copy the key, value to a map<string, string> then put it into the format buffer. I'm wondering whether it will bring performance regression for logging. 🤔

Reply:
spdlog may be a good alternative considering performance. It also has many convenient features.